### PR TITLE
Migrate some kill integration cli test to api tests

### PR DESCRIPTION
**Pull request by vdemeester**
Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -107,7 +107,7 @@ func CheckRedirect(req *http.Request, via []*http.Request) error {
 // Use DOCKER_API_VERSION to set the version of the API to reach, leave empty for latest.
 // Use DOCKER_CERT_PATH to load the TLS certificates from.
 // Use DOCKER_TLS_VERIFY to enable or disable TLS verification, off by default.
-func NewEnvClient() (*Client, error) {
+func NewEnvClient(ops ...func(*Client)) (*Client, error) {
```
Comment (vdemeester) [inline client/client.go]: These changes on `client` are not strictly required (and might be better off discussed in a separate PR)

File: integration-cli/docker_cli_kill_test.go [hunk 2]
```diff
@@ -6,37 +6,9 @@ import (
...
-
-	cli.DockerCmd(c, "stop", cleanedContainerID)
-	cli.WaitExited(c, cleanedContainerID, 10*time.Second)
-
-	cli.Docker(cli.Args("kill", "-s", "30", cleanedContainerID)).Assert(c, icmd.Expected{
-		ExitCode: 1,
-	})
-}
-
 func (s *DockerSuite) TestKillDifferentUserContainer(c *check.C) {
```
Comment (vdemeester) [inline integration-cli/docker_cli_kill_test.go]: This one shouldn't be hard to port either

File: client/client.go (hunk 1 above)
Comment (dnephin) [reply client/client.go]: I think we've wanted to make this change for a while. What do you think about something like this:

* create a `func FromEnv(*Client)` and move all the code from `NewEnvClient` (except the `ops` addition) to `FromEnv()`
* create `NewClientWithOpts()` (or something like that) which creates a client from ops, but has no other inline setup.
* remove `ops`  from `NewEnvClient` and deprecate it. Replace the implementation with `NewClientWithOpts(FromEnv)`.

That way the new interface would create a bare `client{}` and we can apply any ops onto it. If there are required fields (host and client?) maybe it would also validate that.

File: integration/container/kill_test.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+			err = client.ContainerKill(ctx, c.ID, tc.signal)
+			require.NoError(t, err)
+
+			poll.WaitOn(t, containerStatusIs(ctx, client, c.ID, tc.status), poll.WithDelay(100*time.Millisecond))
+		})
+	}
+}
+
+func TestKillWithStopSignalAndRestartPolicies(t *testing.T) {
+	skip.If(t, testEnv.OSType != "linux") // Windows only supports 1.25 or later
```
Comment (dnephin) [inline integration/container/kill_test.go]: Instead of a comment this can be the last arg to `slip.If()` so that we see it in the test output. (same below).

File: client/client.go (hunk 1 above)
Comment (vdemeester) [reply client/client.go]: sounds very good to me 👼

File: integration/container/kill_test.go (hunk 3 above)
Comment (vdemeester) [reply integration/container/kill_test.go]: good point 🙃

File: integration/container/kill_test.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+		},
+		&container.HostConfig{},
+		&network.NetworkingConfig{},
+		"")
+	require.NoError(t, err)
+	err = client.ContainerKill(ctx, c.ID, "SIGKILL")
+	require.NoError(t, err)
+}
+
+func containerStatusIs(ctx context.Context, client client.APIClient, containerID, status string) func(log poll.LogT) poll.Result {
```
Comment (dnephin) [inline integration/container/kill_test.go]: I think this already exists as `containerIsInState` in this package.

**Review by dnephin — COMMENTED** [review]
looks like lint failures, and one comment, otherwise LGTM

**Review by dnephin — APPROVED** [review]
LGTM

File: integration/network/inspect_test.go [hunk 5]
```diff
@@ -24,7 +23,7 @@ func TestInspectNetwork(t *testing.T) {
 	defer setupTest(t)()
 	d := newSwarm(t)
 	defer d.Stop(t)
-	client, err := request.NewClientForHost(d.Sock())
+	client, err := client.NewClientWithOpts(client.WithHost((d.Sock())))
```
Comment (dnephin) [inline integration/network/inspect_test.go]: looks like this test failed

> Cannot connect to the Docker daemon at unix:///tmp/docker-integration/df6a1d02502f7.sock

File: integration/container/kill_test.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/docker/docker/api/types/strslice"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/integration/util/request"
+	"github.com/gotestyourself/gotestyourself/poll"
+	"github.com/gotestyourself/gotestyourself/skip"
+	"github.com/stretchr/testify/require"
+)
+
+func TestKillContainerInvalidSignal(t *testing.T) {
+	defer setupTest(t)()
```
Comment (cpuguy83) [inline integration/container/kill_test.go]: I don't think this works with `t.Parallel()`
For what it's worth, the auto-cleanup is probably not needed as it's simple to cleanup manually.

**Comment by yongtang** [comment]
Overall LGTM. Though powerpc build keeps panicking. Not sure if it is related (or it exposes some other unrelated issue).

File: integration/container/kill_test.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+					}},
+				&network.NetworkingConfig{},
+				"")
+			require.NoError(t, err)
+			err = client.ContainerStart(ctx, c.ID, types.ContainerStartOptions{})
+			require.NoError(t, err)
+			err = client.ContainerKill(ctx, c.ID, "TERM")
+			require.NoError(t, err)
+
+			poll.WaitOn(t, containerIsInState(ctx, client, c.ID, tc.status), poll.WithDelay(100*time.Millisecond))
```
Comment (cpuguy83) [inline integration/container/kill_test.go]: Just a performance concern here.. could this start up a "wait" goroutine earlier on for the case of it being stopped? (side thought, would be nice if the API let us wait for a container to be running).

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
Seems mostly ok, the only thing is I don't think we can use `defer setupTest(t)()` with `t.Parallel()`... and I think we should opt for `t.Parallel()` over the cleanup helper.

File: integration/container/kill_test.go (hunk 7 above)
Comment (cpuguy83) [reply integration/container/kill_test.go]: Funny, just found https://github.com/moby/moby/pull/36147

File: integration/container/kill_test.go (hunk 6 above)
Comment (dnephin) [reply integration/container/kill_test.go]: Good catch. Altough we don't actually use `-parallel` right now so these `t.Parallel()` don't do anything I believe.

Comment (cpuguy83) [reply integration/container/kill_test.go]: I think we should make the extra effort to have these run parallel correctly, as we move stuff over it will be a huge help to test run times.

Comment (yongtang) [reply integration/container/kill_test.go]: While trying to migrate some integration-cli tests to integration api tests, I noticed some of the tests rely on fixed "name" of containers, like `foo`, `server`, etc.. I think that may not work well with `parallel`. May need to convert them to random generated names to avoid collision.

Comment (dnephin) [reply integration/container/kill_test.go]: Ya, I'm not against trying to get things running in parallel, but it is a lot more work than just adding `t.Parallel()`. 

It's a lot easier to get the cleanup wrong if we don't have an easy way to label everything with the test name.

Comment (thaJeztah) [reply integration/container/kill_test.go]: > May need to convert them to random generated names to avoid collision.

Where possible we should just capture the container ID that's returned

Comment (cpuguy83) [reply integration/container/kill_test.go]: Yeah, I think we'd need to have the client automatically namespace objects or something then the defer could cleanup just the namespaces resources.

**Comment by vdemeester** [comment]
@cpuguy83 updated by removing `t.Parallel()` for now.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by yongtang** [comment]
@vdemeester I think there was an usage of `NewClientForHost` in secret_test.go which causes the build failure. Can you take a look?

**Comment by vdemeester** [comment]
@yongtang yep, I will 😉

**Review by yongtang — APPROVED** [review]
LGTM
